File: Programming fundamentals/Associative Arrays/LegendaryFarming.js
function solve(input) {
    const collected = {
        shards: 0,
        fragments: 0,
        motes: 0
    };
    const junk = {};
    const materials = input.split(' ');
    for (let i = 0; i < materials.length; i += 2) {
        const currentMaterial = materials[i + 1].toLowerCase();
        const quantity = Number(materials[i]);
        if (currentMaterial === 'shards' || currentMaterial === 'fragments' || currentMaterial === 'motes') {
            collected[currentMaterial] += quantity;
            if (collected[currentMaterial] >= 250) {
                let forgedItem = '';
                switch (currentMaterial) {
                    case 'shards':
                        forgedItem = 'Shadowmourne';
                        break;
                    case 'fragments':
                        forgedItem = 'Valanyr';
                        break;
                    case 'motes':
                        forgedItem = 'Dragonwrath';
                        break;
                }
                console.log(`${forgedItem} obtained!`);
                collected[currentMaterial] -= 250;
                break;
            }
        } else {
            if (!junk.hasOwnProperty(currentMaterial)) {
                junk[currentMaterial] = 0;
            }
            junk[currentMaterial] += quantity;
        }
    }
    Object.entries(collected)
        .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
        .forEach(kvp => console.log(`${kvp[0]}: ${kvp[1]}`));
    Object.entries(junk)
        .sort((a, b) => a[0].localeCompare(b[0]))
        .forEach(kvp => console.log(`${kvp[0]}: ${kvp[1]}`))
}

solve(
    '3 Motes 5 stones 5 Shards 6 leathers 255 fragments 7 Shards'
)

File: Java Advanced/SetsAndMaps/LegendaryFarming.java
package SetsAndMaps;

import java.util.Comparator;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

public class LegendaryFarming {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        TreeMap<String, Integer> legendaryLoot = new TreeMap<>();
        TreeMap<String, Integer> trashLoot = new TreeMap<>();
        legendaryLoot.put("shards", 0);
        legendaryLoot.put("fragments", 0);
        legendaryLoot.put("motes", 0);

        while (legendaryLoot.get("shards") < 250 &&
                legendaryLoot.get("fragments") < 250 &&
                legendaryLoot.get("motes") < 250) {
            String[] input = scanner.nextLine().split("\\s+");
            for (int i = 0; i < input.length; i += 2) {
                int quantity = Integer.parseInt(input[i]);
                String loot = input[i + 1].toLowerCase();
                if (isLegendary(loot)) {
                    legendaryLoot.putIfAbsent(loot, 0);
                    legendaryLoot.put(loot, legendaryLoot.get(loot) + quantity);
                } else {
                    trashLoot.putIfAbsent(loot, 0);
                    trashLoot.put(loot, trashLoot.get(loot) + quantity);
                }
                if (legendaryLoot.get("shards") >= 250 || legendaryLoot.get("fragments") >= 250 || legendaryLoot.get("motes") >= 250) {
                    break;
                }
            }
        }
        if (legendaryLoot.get("shards") >= 250) {
            System.out.println("Shadowmourne obtained!");
            legendaryLoot.put("shards", legendaryLoot.get("shards") - 250);
        } else if (legendaryLoot.get("fragments") >= 250) {
            System.out.println("Valanyr obtained!");
            legendaryLoot.put("fragments", legendaryLoot.get("fragments") - 250);
        } else if (legendaryLoot.get("motes") >= 250) {
            System.out.println("Dragonwrath obtained!");
            legendaryLoot.put("motes", legendaryLoot.get("motes") - 250);
        }
        legendaryLoot.entrySet().stream()
                .sorted((l1, l2) -> l2.getValue().compareTo(l1.getValue()))
                .forEach(l -> System.out.println(l.getKey() + ": " + l.getValue()));
        trashLoot.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(l -> System.out.println(l.getKey() + ": " + l.getValue()));
    }

    public static boolean isLegendary(String loot) {
        return loot.equals("shards") || loot.equals("fragments") || loot.equals("motes");
    }
}
